Frameworks/QuartzCore/CALayer.mm

```diff
@@ -427,14 +427,15 @@ - (void)renderInContext:(CGContextRef)ctx {
             [priv->delegate drawLayer:self inContext:ctx];
         }
     } else {
-        CGRect rect;
-
-        rect.origin.x = 0;
-        rect.origin.y = priv->bounds.size.height * priv->contentsScale;
-        rect.size.width = priv->bounds.size.width * priv->contentsScale;
-        rect.size.height = -priv->bounds.size.height * priv->contentsScale;
-
-        _CGContextDrawImageRect(ctx, priv->contents, rect, destRect);
+        // If the layer has cached contents, blit them directly.
+
+        // |1  0 0| is the transformation matrix for flipping a rect anchored at 0,0 about its Y midpoint.
+        // |0 -1 0|
+        // |0  h 1|
+        CGContextSaveGState(ctx);
```

Comment:
is saving/popping g-state relatively free?  we should be able to get away with just reverting the CTM if not.